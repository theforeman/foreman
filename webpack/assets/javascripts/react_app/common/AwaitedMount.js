import PropTypes from 'prop-types';
import { useState, useEffect, useCallback } from 'react';
import store from '../redux';
import componentRegistry from '../components/componentRegistry';

// Mounts a component after all plugins have been imported to make sure that all plugins are available to the component
export const AwaitedMount = ({ component, data, flattenData }) => {
  const [mounted, setMounted] = useState(false);
  const [mountedComponent, setMountedComponent] = useState(null);
  const mountComponent = useCallback(async () => {
    if (componentRegistry.registry[component]) {
      setMounted(true);
      setMountedComponent(
        componentRegistry.markup(component, {
          data,
          store,
          flattenData,
        })
      );
    } else {
      // eslint-disable-next-line no-console
      console.debug(
        `Component not found: ${component}. The script for the component might not have been loaded yet.`
      );
    }
  }, [component, data, flattenData]);
  useEffect(() => {
    document.addEventListener('loadPlugin', mountComponent);
    return () => {
      window.removeEventListener('loadPlugin', mountComponent);
    };
  }, [mountComponent]);
  useEffect(() => {
    mountComponent();
  }, [mountComponent]);
  return mounted ? mountedComponent : null;
};

AwaitedMount.propTypes = {
  component: PropTypes.string.isRequired,
  data: PropTypes.object,
  flattenData: PropTypes.bool,
};
AwaitedMount.defaultProps = {
  data: {},
  flattenData: false,
};
