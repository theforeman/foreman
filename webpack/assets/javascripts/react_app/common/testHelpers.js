import React from 'react';
import { shallow } from '@theforeman/test';

jest.useFakeTimers();

export default {
  mockStorage: () => {
    const storage = {};

    return {
      setItem: (key, value) => {
        storage[key] = value || '';
      },
      getItem: key => (key in storage ? storage[key] : null),
      removeItem: key => {
        delete storage[key];
      },
      get length() {
        return Object.keys(storage).length;
      },
      key: i => {
        const keys = Object.keys(storage);

        return keys[i] || null;
      },
    };
  },
};

export const mockWindowLocation = ({ href }) => {
  let currentHref = href;
  delete global.window.location;
  global.window.location = { reload: jest.fn() };
  Object.defineProperty(global.window.location, 'href', {
    configurable: true,
    get: () => currentHref,
    set: newValue => {
      currentHref = newValue;
    },
  });
  return jest.spyOn(global.window.location, 'href', 'set');
};

// a helper method for invoking a class method (for unit tests)
// obj = a class
// func = a tested function
// objThis = an object's this
// arg = function args

export const classFunctionUnitTest = (obj, func, objThis, args) =>
  obj.prototype[func].apply(objThis, args);

/**
 * Shallow render a component multipile times with fixtures
 * @param  {ReactComponent} Component Component to shallow-render
 * @param  {Object}         fixtures  key=fixture description, value=props to apply
 * @return {Object}                   key=fixture description, value=shallow-rendered component
 */
export const shallowRenderComponentWithFixtures = (Component, fixtures) =>
  Object.entries(fixtures).map(([description, props]) => ({
    description,
    component: shallow(<Component {...props} />),
  }));

/**
 * Test a component with fixtures and snapshots
 * @param  {ReactComponent} Component Component to test
 * @param  {Object}         fixtures  key=fixture description, value=props to apply
 */
export const testComponentSnapshotsWithFixtures = (Component, fixtures) =>
  shallowRenderComponentWithFixtures(
    Component,
    fixtures
  ).forEach(({ description, component }) =>
    it(description, () => expect(component).toMatchSnapshot())
  );

const resolveDispatch = async (action, depth) => {
  // if it is async action and we are allowed to go deeper
  if (depth && typeof action === 'function') {
    const dispatch = jest.fn();
    await action(dispatch);
    jest.runOnlyPendingTimers();

    return Promise.all(
      dispatch.mock.calls.map(call => resolveDispatch(call[0], depth - 1))
    );
  }
  // else return the action itself
  return action;
};

/**
 * run an action (sync or async) and returns a call tree
 * @param  {Function}  runAction  Action runner function
 * @param  {Number} states the depth of dispatch calls
 * @return calls result tree to the given depth - array for each branch of calls
 */
export const runActionInDepth = (runAction, depth = 1) =>
  resolveDispatch(runAction(), depth);

/**
 * run an action (sync or async) and except the results to much snapshot
 * @param  {Function}  runAction  Action runner function
 * @return {Promise}
 */
export const testActionSnapshot = async runAction => {
  const actionResults = runAction();

  // if it's an async action
  if (typeof actionResults === 'function') {
    const dispatch = jest.fn();
    await actionResults(dispatch);

    expect(dispatch.mock.calls).toMatchSnapshot();
  } else {
    expect(actionResults).toMatchSnapshot();
  }
};

/**
 * Test actions with fixtures and snapshots
 * @param  {Object} fixtures key=fixture description, value=action runner function
 */
export const testActionSnapshotWithFixtures = fixtures =>
  Object.entries(fixtures).forEach(([description, runAction]) =>
    it(description, () => testActionSnapshot(runAction))
  );

/**
 * Test a reducer with fixtures and snapshots
 * @param  {Function} reducer  reducer to test
 * @param  {Object}   fixtures key=fixture description, value=props to apply
 */
export const testReducerSnapshotWithFixtures = (reducer, fixtures) => {
  const reduce = ({ state, action = {} } = {}) => reducer(state, action);
  Object.entries(fixtures).forEach(([description, action]) =>
    it(description, () => expect(reduce(action)).toMatchSnapshot())
  );
};

/**
 * Test selectors with fixtures and snapshots
 * @param  {Object} fixtures  key=fixture description,
 *                            value=selector runner function
 */
export const testSelectorsSnapshotWithFixtures = fixtures =>
  Object.entries(fixtures).forEach(([description, selectorRunner]) =>
    it(description, () => expect(selectorRunner()).toMatchSnapshot())
  );
