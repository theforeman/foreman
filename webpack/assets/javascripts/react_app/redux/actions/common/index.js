import API from '../../../API';

export const ajaxRequestAction = ({
  dispatch,
  requestAction,
  successAction,
  failedAction,
  url,
  item,
}) => {
  dispatch({ type: requestAction, payload: item });
  return API.get(url)
    .then(({ data }) =>
      dispatch({ type: successAction, payload: { ...item, ...data } })
    )
    .catch(error => dispatch({ type: failedAction, payload: { error, item } }));
};
