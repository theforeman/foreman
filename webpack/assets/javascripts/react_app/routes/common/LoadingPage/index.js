import React from 'react';
import { Spinner } from 'patternfly-react';
import './loadingpage.scss';

const LoadingPage = () => (
  <div id="loading-page">
    <Spinner loading size="lg" />
  </div>
);

export default LoadingPage;
