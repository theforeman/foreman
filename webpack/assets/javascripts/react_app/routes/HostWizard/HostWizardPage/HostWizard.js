import React from 'react';
import { Wizard, Button, Icon } from 'patternfly-react';
import PageLayout from '../../common/PageLayout/PageLayout';
import { translate as __ } from '../../../common/I18n';

class LoadingWizardExample extends React.Component {
  constructor(props) {
    super(props);
    this.state = { showModal: false };
  }
  close = () => {
    this.setState({ showModal: false });
  };
  open = () => {
    this.setState({ showModal: true });
  };
  render() {
    const { showModal } = this.state;

    return (
      <PageLayout header={__('Host Wizard')} searchable={false}>
        <Button bsStyle="primary" bsSize="large" onClick={this.open}>
          Launch loading wizard
        </Button>

        <Wizard show={showModal} onHide={this.close}>
          <Wizard.Header onClose={this.close} title="Wizard Title" />
          <Wizard.Body>
            <Wizard.Row>
              <Wizard.Main>
                <div className="blank-slate-pf">
                  <div
                    className="spinner spinner-lg blank-slate-pf-icon"
                    style={{ float: 'none' }}
                  />
                  <h3 className="blank-slate-pf-main-action">
                    This will be one of our coolest pages.
                  </h3>
                </div>
              </Wizard.Main>
            </Wizard.Row>
          </Wizard.Body>
          <Wizard.Footer>
            <Button
              bsStyle="default"
              className="btn-cancel"
              onClick={this.close}
            >
              Cancel
            </Button>
            <Button bsStyle="default" disabled>
              <Icon type="fa" name="angle-left" />
              Back
            </Button>
            <Button bsStyle="primary" disabled>
              Next
              <Icon type="fa" name="angle-right" />
            </Button>
          </Wizard.Footer>
        </Wizard>
      </PageLayout>
    );
  }
}

export default LoadingWizardExample;
