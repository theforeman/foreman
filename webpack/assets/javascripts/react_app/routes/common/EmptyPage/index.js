import React from 'react';
import PropTypes from 'prop-types';
import { translate as __ } from '../../../common/I18n';
import DefaultEmptyState from '../../../components/common/EmptyState';
import './emptypage.scss';

const EmptyPage = ({ message: { type, text } }) => (
  <DefaultEmptyState
    icon={type === 'error' ? 'error-circle-o' : 'add-circle-o'}
    header={type === 'error' ? __('Error') : __('No Results')}
    description={text}
  />
);

EmptyPage.propTypes = {
  message: PropTypes.shape({
    type: PropTypes.oneOf(['empty', 'error']),
    text: PropTypes.string,
  }),
};

EmptyPage.defaultProps = {
  message: PropTypes.shape({
    type: 'empty',
    text: 'No Results',
  }),
};

export default EmptyPage;
