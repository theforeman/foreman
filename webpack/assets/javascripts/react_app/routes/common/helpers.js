import { foremanUrl, urlBuilder } from '../../common/urlHelpers';

export const getControllerSearchProps = (
  controller,
  id = 'searchBar',
  canCreate = true
) => ({
  controller,
  autocomplete: {
    id,
    searchQuery: '',
    url: urlBuilder(controller, 'auto_complete_search'),
    useKeyShortcuts: true,
  },
  bookmarks: {
    url: foremanUrl('/api/bookmarks'),
    canCreate,
    documentationUrl: `4.1.5Searching`,
  },
});
