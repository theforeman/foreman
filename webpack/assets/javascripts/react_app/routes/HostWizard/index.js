import React from 'react';
import HostWizard from './HostWizardPage';

export default {
  path: '/host_wizard',
  render: props => <HostWizard {...props} />,
};
