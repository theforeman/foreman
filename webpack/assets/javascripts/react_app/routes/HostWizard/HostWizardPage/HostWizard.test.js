import HostWizard from './HostWizard';
import { testComponentSnapshotsWithFixtures } from '../../../common/testHelpers';

const props = {};
const fixtures = {
  'renders LoginPage': props,
};

describe('HostWizard', () => {
  describe('rendering', () => {
    testComponentSnapshotsWithFixtures(HostWizard, fixtures);
  });
});
