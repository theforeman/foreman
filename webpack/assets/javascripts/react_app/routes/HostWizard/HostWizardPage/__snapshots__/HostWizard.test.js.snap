// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HostWizard rendering renders LoginPage 1`] = `
<PageLayout
  beforeToolbarComponent={null}
  breadcrumbOptions={null}
  customBreadcrumbs={null}
  header="Host Wizard"
  isLoading={false}
  onBookmarkClick={[Function]}
  onSearch={[Function]}
  searchProps={Object {}}
  searchQuery=""
  searchable={false}
  toastNotifications={null}
  toolbarButtons={null}
>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsSize="large"
    bsStyle="primary"
    disabled={false}
    onClick={[Function]}
  >
    Launch loading wizard
  </Button>
  <Wizard
    className=""
    dialogClassName=""
    onClose={[Function]}
    onExited={[Function]}
    onHide={[Function]}
    show={false}
  >
    <WizardHeader
      onClose={[Function]}
      title="Wizard Title"
    />
    <WizardBody
      className=""
    >
      <WizardRow
        className=""
      >
        <WizardMain
          className=""
        >
          <div
            className="blank-slate-pf"
          >
            <div
              className="spinner spinner-lg blank-slate-pf-icon"
              style={
                Object {
                  "float": "none",
                }
              }
            />
            <h3
              className="blank-slate-pf-main-action"
            >
              This will be one of our coolest pages.
            </h3>
          </div>
        </WizardMain>
      </WizardRow>
    </WizardBody>
    <WizardFooter
      className=""
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        className="btn-cancel"
        disabled={false}
        onClick={[Function]}
      >
        Cancel
      </Button>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="default"
        disabled={true}
      >
        <Icon
          name="angle-left"
          type="fa"
        />
        Back
      </Button>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsStyle="primary"
        disabled={true}
      >
        Next
        <Icon
          name="angle-right"
          type="fa"
        />
      </Button>
    </WizardFooter>
  </Wizard>
</PageLayout>
`;
